### PR TITLE
Plt 8244 multiple tokens per role

### PR DESCRIPTION
**Pull request by jhbertra**
- [x] Updated transaction API to support multiple tokens per role
- [x] Moved thread token name to top-level create tx API
- [x] Updated constraints to support multiple tokens per role
- [x] Updated REST API to support multiple tokens per role
- [x] Added new integration tests to cover multiple role tokens.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [ ] [Test report is updated](https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe/test/test-report.md) (if relevant)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
    - [ ] Operables are updated with changes to executable command line options.
    - [ ] Deploy charts updated with changes to operables.
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
        - Review required
        - [x] Substantial changes to code, test, or documentation
        - [ ] Change made to Marlowe validator (@bwbush and @palas must be included as reviewers)
        - Review not required
        - [ ] Minor changes to non-critical code, documentation, nix derivations, configuration files, or scripts
        - [ ] Formatting, spelling, grammar, or reorganization
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe-chain-sync/gen/Language/Marlowe/Runtime/ChainSync/Gen.hs [hunk 1]
```diff
@@ -163,7 +163,7 @@ instance Arbitrary PolicyId where
   arbitrary = PolicyId . unScriptHash <$> arbitrary
 
 instance Arbitrary TokenName where
-  arbitrary = TokenName <$> genBytes
+  arbitrary = TokenName . BS.take 32 <$> genBytes
```
Comment (bwbush) [inline marlowe-chain-sync/gen/Language/Marlowe/Runtime/ChainSync/Gen.hs:166]: Someday I'd like to separate our `Arbitrary` instances into ones that are unconstrained versus ones that have ledger or other constraints.

File: marlowe-chain-sync/gen/Language/Marlowe/Runtime/ChainSync/Gen.hs [hunk 2]
```diff
@@ -163,7 +163,7 @@ instance Arbitrary PolicyId where
...
   shrink = genericShrink
 
 instance Arbitrary Datum where
```
Comment (bwbush) [inline marlowe-chain-sync/gen/Language/Marlowe/Runtime/ChainSync/Gen.hs:169]: FYI, the PLT-8148 branch fixes this long-broken `Arbitrary` instance that was able to produce constructors with negative indices. The new Alonzo library now catch this and cause our tests to fail.

File: marlowe-runtime-web/.golden/OpenApi/golden [hunk 3]
```diff
@@ -2186,6 +2189,26 @@
...
                 "oneOf": [
+                    {
+                        "properties": {
+                            "metadata": {
+                                "$ref": "#/components/schemas/TokenMetadata"
+                            },
+                            "recipients": {
+                                "additionalProperties": {
+                                    "format": "int64",
+                                    "maximum": 1.8446744073709551615e19,
```
Comment (bwbush) [inline marlowe-runtime-web/.golden/OpenApi/golden:2200]: Actually, the ledger may not allow a number of tokens this large. We might want to limit it to what the ledger CDDL actual provides.

File: marlowe-runtime/tx/Language/Marlowe/Runtime/Transaction/Constraints.hs [hunk 4]
```diff
@@ -127,37 +122,126 @@ deriving instance Eq (TxConstraints era 'V1)
...
+pattern MintRoleTokens out wit dist <- NEMintRoleTokens out wit (toDistribution -> dist)
+  where
+    MintRoleTokens out wit dist =
+      maybe RoleTokenConstraintsNone (NEMintRoleTokens out wit) $ nonEmptyDistribution dist
+
+pattern SpendRoleTokens :: Set Chain.AssetId -> RoleTokenConstraints era
+pattern SpendRoleTokens roles <- NESpendRoleTokens (NESet.toSet -> roles)
+  where
+    SpendRoleTokens roles =
+      maybe RoleTokenConstraintsNone NESpendRoleTokens $ NESet.nonEmptySet roles
```
Comment (bwbush) [inline marlowe-runtime/tx/Language/Marlowe/Runtime/Transaction/Constraints.hs:152]: Nice!

**Review by bwbush — APPROVED** [review]
Looks great. Thanks!

File: marlowe-chain-sync/gen/Language/Marlowe/Runtime/ChainSync/Gen.hs (hunk 1 above)
Comment (jhbertra) [reply marlowe-chain-sync/gen/Language/Marlowe/Runtime/ChainSync/Gen.hs:166]: Ideally, our types should enforce ledger constraints where appropriate via smart constructors.

File: marlowe-runtime-web/.golden/OpenApi/golden (hunk 3 above)
Comment (jhbertra) [reply marlowe-runtime-web/.golden/OpenApi/golden:2200]: I think this is something that needs to be applied to the REST API types as a whole - basically, they are unwrapped raw data from `cardano-api`.

**Pull request merged**